// @wapps/storybook-addon-state/lib/Store.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Store {
  constructor(state = {}) {
    this.state = _objectSpread({}, state);
    this.handlers = []; // observers
  }

  set(newState) {
    this.state = _objectSpread({}, this.state, newState);
    this.fire();
  }

  get(key) {
    return this.state[key];
  }

  subscribe(fn) {
    this.handlers.push(fn);
    return fn;
  }

  unsubscribe(fn) {
    this.handlers = this.handlers.filter(handler => handler !== fn);
  }

  fire() {
    this.handlers.forEach(handler => handler(this.state));
  }

}

var _default = Store;
exports.default = _default;
